src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/SyntaxResult.cs

```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Language;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using SyntaxNode = Microsoft.AspNetCore.Razor.Language.Syntax.SyntaxNode;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic
+{
+    internal class SyntaxResult
```

Comment:
Looking at your semantic token info service I'm not sure the value of this class. All it's really doing is ensuring range is calculated once which in practice if it were generated in each method it'd only be generaetd once anyhow. It also captures the `SyntaxKind` and lazily compares it against a specific set of token data types which we care about for semantic meaning.

Is there value to having this separate? It seems like it just adds extra complexity vs. just passing around SyntaxNodes everywhere and computing their ranges/kinds on the fly.

Asking because I feel like the semantic token info service would be a lot more clear without this. But then again it's totally possibly i'm missing something.

Reply:
The problem lies in "computing ranges/kinds on the fly". The nodes that we add to `TagHelperData` aren't of type `MarkupTagHelperDirectiveAttributeSyntax`, they're things like `MarkupTextLiteralSyntax` and `RazorMetaCodeSyntax` because the nodes we store are `node.Transition` and `node.Name`. In order to know which kind we ought to apply once all we have is the node we'd have to inspect that nodes parent and contents in order to understand what it is. I figured it was cleaner to add the `SyntaxResult` POCO than to have logic about our syntax tree bleed out of the `TagHelperSpanVisitor`.

Reply:
I see. I didn't realize you were grabbing pieces of the syntax, storing them in an object and then providing a different kind. Doesn't that couple your TagHelperVisitor impl to your semantic token info service then? Aka, it's building SyntaxResults with pieces of the document and then associating them with a different kind because it knows the semantic token info service is going to classify those pieces a certain way.

I'd suggest taking one of two routes:
1. Make your TagHelper visitor a little more dumb and have it just extract the various TagHelper-like SyntaxNodes. At which point the semantic info service would iterate over those and pull out the pieces it wants to classify. For instance, for a TagHelperDirectiveAttribute it'd extract the colon, name, parameters etc. This would get rid of SyntaxResult.
2. Make the TagHelperVisitor smarter and make it a lot more semantic aware. Instead of returning kinds have it do the SyntaxKind -> SemanticKind (uint) convesrion. This would mean renaming `SyntaxResult` to be more of a SemanticRange where it takes a range and a uint for the semantic token kind.